### PR TITLE
feat: add support for multimodal

### PR DESCRIPTION
**Pull request by leehuwuj**
## Summary by CodeRabbit

- **New Features**
	- Introduced `MultiModalSynthesizer` for enhanced multi-modal query processing, integrating both text and image nodes.
	- Added asynchronous synthesis capabilities with the `asynthesize` method.
	- Enhanced settings to support multi-modal language models based on specific model configurations.

- **Bug Fixes**
	- Improved error handling for model support initialization.

- **Documentation**
	- Updated import statements and method signatures for clarity and maintainability.

### CODE REVIEW

File: templates/components/settings/python/settings.py [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
-from typing import Dict
+from typing import Dict, Optional
 
+from llama_index.core.multi_modal_llms import MultiModalLLM
 from llama_index.core.settings import Settings
 
 
+def get_multi_modal_llm() -> Optional[MultiModalLLM]:
+    model_provider = os.getenv("MODEL_PROVIDER")
+    llm_model = os.getenv("MULTIMODAL_LLM_MODEL")
```
Comment (marcusschiesser) [inline templates/components/settings/python/settings.py:10]: we still should use the model env variable and return null if it's not multimodal

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 2]
```diff
@@ -17,11 +20,18 @@ def create_query_engine(index, **kwargs):
...
     # use auto_routed mode for better query results
-    if (
-        index.__class__.__name__ == "LlamaCloudIndex"
-        and kwargs.get("auto_routed") is None
-    ):
-        kwargs["auto_routed"] = True
+    if index.__class__.__name__ == "LlamaCloudIndex":
+        if kwargs.get("auto_routed") is None:
+            kwargs["auto_routed"] = True
+            kwargs["retrieve_image_nodes"] = True
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:26]: should only be true for multi-modal

File: templates/components/settings/python/settings.py [hunk 3]
```diff
@@ -1,8 +1,12 @@
 import os
-from typing import Dict
+from typing import Dict, Optional
 
+from llama_index.core.multi_modal_llms import MultiModalLLM
 from llama_index.core.settings import Settings
 
+# Singleton for multi-modal LLM
```
Comment (marcusschiesser) [inline templates/components/settings/python/settings.py:7]:
```suggestion
# Singleton for multi-modal LLM (as `Settings` does not support setting `MultiModalLLM`)
```

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 4]
```diff
@@ -17,11 +128,18 @@ def create_query_engine(index, **kwargs):
...
-        index.__class__.__name__ == "LlamaCloudIndex"
-        and kwargs.get("auto_routed") is None
-    ):
-        kwargs["auto_routed"] = True
+    if index.__class__.__name__ == "LlamaCloudIndex":
+        retrieval_mode = kwargs.get("retrieval_mode")
+        if retrieval_mode is None:
+            kwargs["retrieval_mode"] = "auto_routed"
+            if multi_modal_llm:
+                # Note: image nodes are not supported for auto_routed or chunk retrieval mode
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:136]: that comment is confusing as you set retrieval_mode to "auto_routed" above

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 5]
```diff
@@ -1,28 +1,106 @@
...
+    ImageNode,
+    NodeWithScore,
+)
 from llama_index.core.tools.query_engine import QueryEngineTool
 
+from app.settings import get_multi_modal_llm
 
-def create_query_engine(index, **kwargs):
+
+class MultiModalSynthesizer(TreeSummarize):
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:26]: use base synthesizer and pass a responsesynthesizer to the constructor. then you can delegate to this synthesizer

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 6]
```diff
@@ -1,28 +1,106 @@
...
+            kwargs["retrieval_mode"] = "auto_routed"
+            multi_modal_llm = get_multi_modal_llm()
+            if multi_modal_llm:
+                kwargs["retrieve_image_nodes"] = True
+                return RetrieverQueryEngine(
+                    retriever=index.as_retriever(**kwargs),
+                    response_synthesizer=MultiModalSynthesizer(
+                        multimodal_model=multi_modal_llm
+                    ),
+                )
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:101]: how about using `MultiModalSynthesizer` also for non llamacloudindex if there's a mm LLM?

File: templates/components/engines/python/agent/tools/query_engine.py [hunk 7]
```diff
@@ -1,28 +1,106 @@
...
+
+    raise ValueError("Multi-modal LLM is not set")
```
Comment (marcusschiesser) [inline templates/components/engines/python/agent/tools/query_engine.py:103]: should also work without MM LLM

Comment (leehuwuj) [reply templates/components/engines/python/agent/tools/query_engine.py:103]: Ah, i added this just for testing, let me remove it :D

File: templates/components/engines/python/agent/tools/query_engine.py (hunk 6 above)
Comment (leehuwuj) [reply templates/components/engines/python/agent/tools/query_engine.py]: Yes, it makes sense. The `MultiModalSynthesizer` should fallback to the default synthesized if there is no image nodes then we still have the default behavior

**Pull request merged**